Sources/MapboxMaps/Gestures/GestureHandlers/PanGestureHandler.swift

```diff
@@ -33,38 +39,63 @@ internal final class PanGestureHandler: GestureHandler {
             cameraAnimationsManager.cancelAnimations()
             delegate?.gestureBegan(for: .pan)
         case .changed:
-            guard let initialTouchLocation = initialTouchLocation,
-                  let initialCameraState = initialCameraState,
-                  let panScrollingMode = delegate?.panScrollingMode else {
-                return
-            }
+            lastChangedDate = dateProvider.now
             cameraAnimationsManager.cancelAnimations()
-
-            // Reset the camera to its state when the gesture began
-            mapboxMap.setCamera(to: CameraOptions(cameraState: initialCameraState))
-
-            let clampedTouchLocation: CGPoint
-            switch panScrollingMode {
-            case .horizontal:
-                clampedTouchLocation = CGPoint(x: touchLocation.x, y: initialTouchLocation.y)
-            case .vertical:
-                clampedTouchLocation = CGPoint(x: initialTouchLocation.x, y: touchLocation.y)
-            case .horizontalAndVertical:
-                clampedTouchLocation = touchLocation
+            handleChange(withTouchLocation: touchLocation)
+        case .ended:
+            // decelerate
+            guard let lastChangedDate = lastChangedDate,
+                  // if it's been more than 2 frames at 60 Hz since the last change, don't drift
+                  dateProvider.now.timeIntervalSince(lastChangedDate) < 2.0 / 60.0,
+                  let decelerationRate = delegate?.decelerationRate else {
+                return
             }
-
-            // Execute the drag relative to the initial touch location
-            mapboxMap.dragStart(for: initialTouchLocation)
-            let dragCameraOptions = mapboxMap.dragCameraOptions(
-                from: initialTouchLocation,
-                to: clampedTouchLocation)
-            mapboxMap.setCamera(to: dragCameraOptions)
-            mapboxMap.dragEnd()
-        case .ended, .cancelled:
+            cameraAnimationsManager.decelerate(
+                location: touchLocation,
+                velocity: gestureRecognizer.velocity(in: view),
+                decelerationRate: decelerationRate,
+                locationChangeHandler: handleChange(withTouchLocation:),
+                completion: {
+                    self.initialTouchLocation = nil
+                    self.initialCameraState = nil
+                    self.lastChangedDate = nil
+                })
```

Comment:
Does setting all these values to `nil` need to be done? I see that we assign these values when the pan gesture starts.

Reply:
Probably not strictly necessary. It's a bit of defensive programming b/c it ensures that if you enter .changed without having entered .began, we'll just ignore it. Will that ever actually happen? Probably not.

That said, we'll ultimately need to capture self in this completion block to tell the delegate when the deceleration is complete, and there's no danger in letting that be a strong ref if you work out the dependency graph.

Reply:
Gotcha, thanks for the info.